codegenerator/cli/templates/static/codegen/src/eventFetching/SourceManager.res

```diff
@@ -0,0 +1,176 @@
+  ~allPartitions: PartitionedFetchState.allPartitions,
+  ~currentBlockHeight,
+  ~executePartitionQuery,
+  ~waitForNewBlock,
+  ~onNewBlock,
+  ~maxPerChainQueueSize,
+  ~setMergedPartitions,
+  ~stateId,
+) => {
+  if stateId < sourceManager.currentStateId {
+    ()
+  } else {
+    if stateId != sourceManager.currentStateId {
+      sourceManager.currentStateId = stateId
+
+      // Reset instead of clear, so updating state from partitions from prev state doesn't corrupt data
+      sourceManager.allPartitionsFetchingState = []
+    }
+    let {logger, allPartitionsFetchingState, maxPartitionConcurrency} = sourceManager
+
+    let fetchingPartitions = Utils.Set.make()
+    // Js.Array2.forEachi automatically skips empty items
+    allPartitionsFetchingState->Js.Array2.forEachi(({isFetching}, partitionId) => {
+      if isFetching {
+        let _ = fetchingPartitions->Utils.Set.add(partitionId)
+      }
+    })
+    let readyPartitions =
+      allPartitions->PartitionedFetchState.getReadyPartitions(
+        ~maxPerChainQueueSize,
+        ~fetchingPartitions,
+      )
+
+    let mergedPartitions = Js.Dict.empty()
+    let hasQueryWaitingForNewBlock = ref(false)
+    let queries = readyPartitions->Array.keepMap(({fetchState, partitionId}) => {
+      let mergedFetchState = fetchState->FetchState.mergeRegistersBeforeNextQuery
+      if mergedFetchState !== fetchState {
+        mergedPartitions->Js.Dict.set(partitionId->(Utils.magic: int => string), mergedFetchState)
+      }
+      switch mergedFetchState->FetchState.getNextQuery(~partitionId) {
+      | Done => None
+      | NextQuery(nextQuery) =>
+        switch allPartitionsFetchingState->Belt.Array.get(partitionId) {
+        // Deduplicate queries when fetchBatch is called after
+        // isFetching was set to false, but state isn't updated with fetched data
+        | Some({lastFetchedQueryId}) if lastFetchedQueryId === toQueryId(nextQuery) => None
+        | _ => {
+            let {fromBlock, toBlock} = nextQuery
+            if fromBlock > currentBlockHeight {
+              hasQueryWaitingForNewBlock := true
+              None
+            } else {
+              switch toBlock {
+              | Some(toBlock) if toBlock < fromBlock =>
+                //This is an invalid case. We should never arrive at this match arm but it would be
+                //detrimental if it were the case.
+                FromBlockIsHigherThanToBlock({fromBlock, toBlock})->ErrorHandling.mkLogAndRaise(
+                  ~logger,
+                  ~msg="Unexpected error getting next query in partition",
+                )
+              | _ => ()
+              }
+              Some(nextQuery)
+            }
+          }
+        }
+      }
+    })
+    setMergedPartitions(mergedPartitions)
+
+    switch (queries, currentBlockHeight) {
+    | ([], _)
+    | // For the case with currentBlockHeight=0 we should
+    // force getting the known chain block, even if there are no ready queries
```

Comment:
Does this handle a chain with endblock?

Reply:
Yes. It filters out all completely fetched partitions